test/library/encryption/perf_helpers.go

```diff
@@ -0,0 +1,193 @@
+
+	statsStore map[string]int
+	lockStats  *sync.Mutex
+	wg         *sync.WaitGroup
+}
+
+func newRunner() *runner {
+	r := &runner{}
+
+	r.errorStore = map[string]int{}
+	r.lock = &sync.Mutex{}
+	r.statsStore = map[string]int{}
+	r.lockStats = &sync.Mutex{}
+
+	r.wg = &sync.WaitGroup{}
+
+	return r
+}
+
+func (r *runner) run(t testing.TB, workers int, workFunc ...DBLoaderFuncType) {
+	t.Logf("Executing provided load function for %d workers", workers)
+	for i := 0; i < workers; i++ {
+		wrapper := func(wg *sync.WaitGroup) {
+			defer wg.Done()
+			kubeClient, err := newKubeClient(300, 600)
+			if err != nil {
+				t.Errorf("Unable to create a kube client for a worker due to %v", err)
+				r.collectError(err)
+				return
+			}
+			_ = runWorkFunctions(kubeClient, "", r.collectError, r.collectStat, workFunc...)
+		}
+		r.wg.Add(1)
+		go wrapper(r.wg)
+	}
+	r.wg.Wait()
+	t.Log("All workers completed successfully")
+}
+
+func (r *runner) collectError(err error) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	errCount, ok := r.errorStore[err.Error()]
+	if !ok {
+		r.errorStore[err.Error()] = 1
+		return
+	}
+	errCount += 1
+	r.errorStore[err.Error()] = errCount
+}
+
+func (r *runner) collectStat(stat string) {
+	r.lockStats.Lock()
+	defer r.lockStats.Unlock()
+	statCount, ok := r.statsStore[stat]
+	if !ok {
+		r.statsStore[stat] = 1
+		return
+	}
+	statCount += 1
+	r.statsStore[stat] = statCount
+}
+
+func runWorkFunctions(kubeClient kubernetes.Interface, namespace string, errorCollector func(error), statsCollector func(string), workFunc ...DBLoaderFuncType) error {
+	if len(namespace) == 0 {
+		namespace = createNamespaceName()
+	}
+	for _, work := range workFunc {
+		err := work(kubeClient, namespace, errorCollector, statsCollector)
+		if err != nil {
+			errorCollector(err)
+			return err
+		}
+	}
+	return nil
+}
+
+func DBLoaderRepeat(times int, genNamespaceName bool, workToRepeatFunc ...DBLoaderFuncType) DBLoaderFuncType {
+	return DBLoaderRepeatParallel(times, 1, genNamespaceName, workToRepeatFunc...)
+}
+
+func DBLoaderRepeatParallel(times int, workers int, genNamespaceName bool, workToRepeatFunc ...DBLoaderFuncType) DBLoaderFuncType {
+	return func(kubeClient kubernetes.Interface, namespace string, errorCollector func(error), statsCollector func(string)) error {
+		if times < workers {
+			panic("DBLoaderRepeat cannot be < workers")
+		}
+		wg := &sync.WaitGroup{}
+		workPerWorker := times / workers
+		for w := 0; w < workers; w++ {
+			work := func(wg *sync.WaitGroup) {
+				defer wg.Done()
+				for i := 0; i < workPerWorker; i++ {
+					if genNamespaceName {
+						namespace = createNamespaceName()
+					}
+					err := runWorkFunctions(kubeClient, namespace, errorCollector, statsCollector, workToRepeatFunc...)
+					if err != nil {
+						errorCollector(err)
+					}
+				}
+			}
+			wg.Add(1)
+			go work(wg)
+		}
+		wg.Wait()
+		return nil
+	}
+}
+
+func createNamespaceName() string {
+	return fmt.Sprintf("encryption-%s", rand.String(10))
```

Comment:
is this random enough?

Reply:
yes, the tests we have atm generate `3` namespaces at most :)

Reply:
Yeah, but the chance that two of the 3 hit the same namespace is high https://www.dcode.fr/birthday-problem

Reply:
Does it mean we have a 50% chance of a collision after generating `60466176` namespaces?